pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,423 @@
+package cos
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/mozillazg/go-cos"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HeadObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const dirDelim = "/"
+
+// Bucket implements the store.Bucket interface against cos-compatible(Tencent Object Storage) APIs.
+type Bucket struct {
+	logger   log.Logger
+	client   *cos.Client
+	opsTotal *prometheus.CounterVec
+	name     string
+}
+
+// cosConfig encapsulates the necessary config values to instantiate an cos client.
+type cosConfig struct {
+	Bucket    string `yaml:"bucket"`
+	Region    string `yaml:"region"`
+	AppId     string `yaml:"appid"`
+	SecretKey string `yaml:"secret-key"`
```

Comment:
can we have snake_case here? (:

Reply:
OK, my bad.

Reply:
Created an issue to standardise this #596 across the code base.